starter_project/normalize_breton_lib.py

```diff
@@ -1,19 +1,98 @@
-HYPHEN = "-"
-sigma_star = union(BR_GRAPHEMES, HYPHEN).closure()
+MUTATION_HANDLER = mutation_handler.CelticMutationHandler()
+MUTATION_HANDLER.breton_handler()
 
-DELETE_HYPHEN = transducer(HYPHEN, "")
+GRAPHEMES = MUTATION_HANDLER.graphemes
 
-REMOVE_EXTRA_HYPHEN_IN_WORD = cdrewrite(
-  DELETE_HYPHEN,
-  BR_GRAPHEMES + acceptor(HYPHEN),
-  BR_GRAPHEMES,
-  sigma_star)
+SOFT_TRIGGERS = MUTATION_HANDLER.soft_triggers
+HARD_TRIGGERS = MUTATION_HANDLER.hard_triggers
+SPIRANT_TRIGGERS = MUTATION_HANDLER.spirant_triggers
+MIXED_TRIGGERS = MUTATION_HANDLER.mixed_triggers
+NASAL_TRIGGERS = MUTATION_HANDLER.nasal_triggers
+LENITION_TRIGGERS = MUTATION_HANDLER.lenition_triggers
+ECLIPSIS_TRIGGERS = MUTATION_HANDLER.eclipsis_triggers
 
-def NormalizeBreton(breton_string):
-  return compose(breton_string, REMOVE_EXTRA_HYPHEN_IN_WORD).string()
+SOFT_MUTATION = MUTATION_HANDLER.soft_map
+HARD_MUTATION = MUTATION_HANDLER.hard_map
+SPIRANT_MUTATION = MUTATION_HANDLER.spirant_map
+MIXED_MUTATION = MUTATION_HANDLER.mixed_map
+NASAL_MUTATION = MUTATION_HANDLER.nasal_map
+LENITION_MUTATION = MUTATION_HANDLER.lenition_map
+ECLIPSIS_MUTATION = MUTATION_HANDLER.eclipsis_map
+
+SPACE = " "
+SIGMA_STAR = MUTATION_HANDLER.sigma_star
+
+PREPROCESS = MUTATION_HANDLER.preprocessing
+POSTPROCESS = MUTATION_HANDLER.postprocessing
+
+DO_SOFT_MUTATION = cdrewrite(
+    SOFT_MUTATION,
+    union(SPACE, "[BOS]") + SOFT_TRIGGERS + acceptor(SPACE),
+    GRAPHEMES,
+    SIGMA_STAR)
+DO_HARD_MUTATION = cdrewrite(
+    HARD_MUTATION,
+    union(SPACE, "[BOS]") + HARD_TRIGGERS + acceptor(SPACE),
+    GRAPHEMES,
+    SIGMA_STAR)
+DO_SPIRANT_MUTATION = cdrewrite(
+    SPIRANT_MUTATION,
+    union(SPACE, "[BOS]") + SPIRANT_TRIGGERS + acceptor(SPACE),
+    GRAPHEMES,
+    SIGMA_STAR)
+DO_MIXED_MUTATION = cdrewrite(
+    MIXED_MUTATION,
+    union(SPACE, "[BOS]") + MIXED_TRIGGERS + acceptor(SPACE),
+    GRAPHEMES,
+    SIGMA_STAR)
+DO_NASAL_MUTATION = cdrewrite(
+    NASAL_MUTATION,
+    union(SPACE, "[BOS]") + NASAL_TRIGGERS + acceptor(SPACE),
+    GRAPHEMES,
+    SIGMA_STAR)
+DO_LENITION = cdrewrite(
+    LENITION_MUTATION,
+    union(SPACE, "[BOS]") + LENITION_TRIGGERS + acceptor(SPACE),
+    GRAPHEMES,
+    SIGMA_STAR)
+DO_ECLIPSIS = cdrewrite(
+    ECLIPSIS_MUTATION,
+    union(SPACE, "[BOS]") + ECLIPSIS_TRIGGERS + acceptor(SPACE),
+    GRAPHEMES,
+    SIGMA_STAR)
+
+DO_PREPROCESSING = cdrewrite(
+    PREPROCESS,
+    union(SPACE, "[BOS]"),
+    union(GRAPHEMES, SPACE),
+    SIGMA_STAR)
+
+DO_POSTPROCESSING = cdrewrite(
+    POSTPROCESS,
+    union(SPACE, "[BOS]"),
+    union(GRAPHEMES, SPACE),
+    SIGMA_STAR)
+
+
+def normalize_breton(breton_string: str) -> str:
+    """Applies Breton mutations."""
+    return (breton_string.strip().lower() @
+            DO_PREPROCESSING @
+            DO_SOFT_MUTATION @
+            DO_HARD_MUTATION @
+            DO_SPIRANT_MUTATION @
+            DO_POSTPROCESSING).string()
+
+def normalize_welsh(welsh_string: str) -> str:
```

Comment:
Doesn't matter for this starter project, but I think the design/setup is not quite right here - this file is called normalize_breton_lib, but you have stuff for Welsh in here. Not important for now, but if you return to this, I would try and make the mutation handler as language-agnostic as possible, and put all the language-specific stuff (including the triggers and contexts for Breton that are currently over there) in language-specific config files.